win: map ERROR_ELEVATION_REQUIRED to UV_EACCES

richardlau opened the pull request:
uv_spawn() on Windows will fail with ERROR_ELEVATION_REQUIRED (740) if
attempting to run an application that requires elevation.

~~Add a Windows test that calls uv_spawn() with a helper .exe that
requires elevated permissions. Test should either run successfully if
run with Administrator privileges or fail with the mapped error.~~

Refs: https://github.com/nodejs/node/issues/9464

richardlau commented:
I've only added the new .exe helper to `uv.gyp` -- Do I need to add to `Makefile.am` too (it's Windows only)? If the answer is yes I'll probably need help because the helper needs to be built with `requireAdministrator` (`'UACExecutionLevel': 2,` in the gyp file).

cjihrig commented on test/elevation.c:
```diff
@@ -0,0 +1,3 @@
+int main(int argc, char **argv) {
+    return 0;
```
Two space indentation please.

richardlau replied:
done

cjihrig commented on test/test-spawn.c:
```diff
@@ -215,6 +215,52 @@ TEST_IMPL(spawn_fails_check_for_waitpid_cleanup) {
+  ASSERT(r == 0);
+  exepath[exepath_size] = '\0';
+  snprintf(path, path_size, "%s\\..\\elevation.exe", exepath);
+  uv_loop_t* loop = uv_default_loop();
```
Can you move all variable declarations to the top of the function please.

richardlau replied:
done

richardlau commented:
`UV_EACCES` was picked as the value to map to as it seems to be the closest to what you see on the terminal if running the helper:

With `cmd.exe` (Answering `no` to the UAC pop-up):
```
C:\work\node\github\libuv>Debug\elevation.exe
Access is denied.

C:\work\node\github\libuv>
```
In git-bash (no UAC pop-up):
```
riclau@PK1KZCZ MINGW64 /c/work/node/github/libuv (elevation)
$ Debug/elevation.exe
bash: Debug/elevation.exe: Permission denied

riclau@PK1KZCZ MINGW64 /c/work/node/github/libuv (elevation)
$
```

richardlau commented:
Updated to add license boilerplate to the new file.

santigimeno commented on test/elevation.c:
```diff
@@ -0,0 +1,23 @@
+/* Copyright Joyent, Inc. and other Node contributors. All rights reserved.
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
```
Nowadays the first line usually looks like [this](https://github.com/libuv/libuv/blob/v1.x/test/test-udp-alloc-cb-fail.c#L1)

richardlau replied:
Fair enough, I'll change it. Since different source files have older/newer style boilerplate it might be an idea to update [`CONTRIBUTING.md`](https://github.com/libuv/libuv/blob/v1.x/CONTRIBUTING.md) to point to the correct one as currently all it says is
```
Look at other tests to see how they should be structured (license boilerplate, the way entry points are declared, etc.).
```

santigimeno commented on test/test-spawn.c:
```diff
@@ -215,6 +215,54 @@ TEST_IMPL(spawn_fails_check_for_waitpid_cleanup) {
+  uv_fs_t req;
+  uv_loop_t* loop = uv_default_loop();
+  char path[1024];
+  size_t path_size = 1024;
```
Can you split declaration and assignment? Apparently it produces warnings on older VS

richardlau replied:
Sure, will do.

richardlau replied:
In fact I'll probably get rid of `path_size` and use `sizeof(path)` instead.

santigimeno commented on test/test-spawn.c:
```diff
@@ -215,6 +215,54 @@ TEST_IMPL(spawn_fails_check_for_waitpid_cleanup) {
+
+  printf("Launching %s\n", (char*) req.ptr);
+  r = uv_spawn(uv_default_loop(), &process, &options);
+  printf("r = %d, %s - %s\n", r, uv_err_name(r), uv_strerror(r));
```
I would remove the `printf()`

richardlau replied:
hmm okay. I found it useful for diagnostics. If I comment out the mapping, then with the `printf()` calls the test fails:
```
C:\work\node\github\libuv>Debug\run-tests spawn_requires_elevation
not ok 1 - spawn_requires_elevation
# exit code 3
# Output from process `spawn_requires_elevation`:
# Launching C:\work\node\github\libuv\Debug\elevation.exe
# r = -4094, UNKNOWN - unknown error
# Assertion failed in test\test-spawn.c on line 257: r == 0 || r == UV_EACCES

C:\work\node\github\libuv>
```
Compared to (without `printf()`'s):
```
C:\work\node\github\libuv>Debug\run-tests spawn_requires_elevation
not ok 1 - spawn_requires_elevation
# exit code 3
# Output from process `spawn_requires_elevation`:
# Assertion failed in test\test-spawn.c on line 255: r == 0 || r == UV_EACCES

C:\work\node\github\libuv>
```

santigimeno replied:
Ack

santigimeno commented on test/test-spawn.c:
```diff
@@ -215,6 +215,54 @@ TEST_IMPL(spawn_fails_check_for_waitpid_cleanup) {
+   * Will either succeed or fail depending on whether test is being run with
+   * elevated permissions
+   */
+  ASSERT(r == 0 || r == UV_EACCES);
```
Is there a way to programmatically know whether the process is being run with elevated permissions or not?

santigimeno replied:
Maybe [this](http://stackoverflow.com/a/8196291/1562178) is usefule

richardlau commented:
Updated:
[x] Fixed boilerplate
~~[ ] Split variable declaration and assignment~~ Replaced `path_size` with `sizeof(path)` and `loop` with `uv_default_loop()`
[x] Check if process is being run with elevated permissions and assert accordingly

I'll squash commits later if this looks good.

richardlau commented on test/test-spawn.c:
```diff
@@ -215,6 +215,89 @@ TEST_IMPL(spawn_fails_check_for_waitpid_cleanup) {
+    ASSERT(r == UV_EACCES);
+  }
+  ASSERT(0 == uv_run(uv_default_loop(), UV_RUN_DEFAULT));
+
```
Doesn't look like it. Removed.

santigimeno commented:
LGTM if CI is happy: https://ci.nodejs.org/job/libuv-test-commit/185/

cjihrig commented on test/test-spawn.c:
```diff
@@ -215,6 +215,88 @@ TEST_IMPL(spawn_fails_check_for_waitpid_cleanup) {
+    ASSERT(r == 0);
+    r = uv_read_start((uv_stream_t*) &out, on_alloc, on_read);
+    ASSERT(r == 0);
+    ASSERT(0 == uv_run(uv_default_loop(), UV_RUN_DEFAULT));
```
Can you switch the order here. This is the only one where `0` is on the left hand side.

cjihrig commented on test/test-spawn.c:
```diff
@@ -215,6 +215,88 @@ TEST_IMPL(spawn_fails_check_for_waitpid_cleanup) {
+  BOOL elevated = FALSE;
+  HANDLE h_token = NULL;
+  if (OpenProcessToken(GetCurrentProcess(), TOKEN_QUERY, &h_token)) {
+    TOKEN_ELEVATION elevation_token;
```
Can you move all of the variable declarations to the top of the function please.

richardlau replied:
done

richardlau commented:
Updated with variable declarations moved to top of function. 

I've decided to simplify the `spawn_requires_elevation` test and simply skip the test if the process is already elevated (since the actual change in libuv is mapping the error code when the process is not elevated).

richardlau commented:
Spotted and fixed another ASSERT where the 0 was on the left hand side for consistency.

saghul commented:
Can this be written without an external binary we need to compile now?  That binary would also need to be compiled with autotools for MinGW, which seems to overcomplicate things IMHO.

richardlau commented:
Well we could drop the new test and just rely on the addition to `test-error.c`.

To test using `uv_spawn` we need to identify a binary on Windows that:
a) Requires elevation
b) Is present on all flavors of Windows

saghul commented:
@richardlau 👍 That sounds like a better approach to me.

richardlau commented:
> @richardlau :+1: That sounds like a better approach to me.

Dropping the new test?

saghul commented:
Yeah, dropping the new binary.

richardlau commented:
Updated. New test and binary dropped -- Just the new mapping and addition to `test-error.c`. This is a much simpler pull request now :grin:

saghul commented:
> To test using uv_spawn we need to identify a binary on Windows that:
> a) Requires elevation
> b) Is present on all flavors of Windows

I still wanted this though :-)  Is there no such a binary?

richardlau commented:
I haven't been able to find one so far. All of the built in Windows commands that I have tried so far (`net session`, `fsutil`, `openfiles`) require administrative privilege but fail in a different way -- They print friendly messages out to the console and do not cause uv_spawn to return an error.

richardlau commented:
I've put back the new binary for now, even though it's not ideal. If anyone has other suggestions for binaries to try please feel free to comment or try by replacing the options just before the `uv_spawn` call in the new test.

I'll see if I can get a MinGW environment set up to look at the autotools side.

saghul commented:
I'm -1 on landing the new binary. @libuv/collaborators WDYT?

santigimeno commented:
My take is that I'd rather have a test that none but won't oppose to its removal.
From my ignorance, if having the binary built for `MinGW` is a problem, would disabling the test on `MinGW` an option?

saghul commented:
@santigimeno I don't disagree, but having such a convoluted test just for a errno map is overkill IMHO.

cjihrig commented:
I agree that it seems like overkill if its only use is testing the errno mapping.

richardlau commented:
Based on the most recent comments I've removed the new binary and test again.

cjihrig commented:
Landed in 11ce5df5eadd67fa951b8c4c6293fa784d21cf80. Thanks!
